### PR TITLE
Issue 847/better api loading

### LINKED ISSUE #847: Handle caching in new model/redux design
> In this code, domain data (like `callBackLater`, `target` and `title`) is mixed with loading/caching support meta-data (`isStale, isLoading`) in the same object. I think it should be separated, so that there's a "container" object (defined elsewhere) with the meta-data, and a data field for the actual data, i.e. something like this:

```ts
interface CachedData<T> {
  isLoading: boolean;
  isStale: boolean;
  data: T;
}
```

_Originally posted by @richardolsson in https://github.com/zetkin/app.zetkin.org/pull/845#discussion_r1031359228_

### PR DESCRIPTION
**Pull request by richardolsson**
## Description
This PR refactors the separation of concerns between the existing "model" concept and the new "repository" and API client concepts.

## Screenshots
The image below illustrates the relationships between the new classes.

image

## Changes
* Adds API client classes
* Adds `Environment` class to store API client and redux store and pass it down the React tree using context
* Moves caching/loading logic from `CallAssignmentModel` to new `CallAssignmentsRepo`
* Adds a notion of "futures", an interface which contains data and it's loading state
* Restructures redux store to store caching information adjacent to the data retrieved from the API
* Adds tests for `CallAssignmentsModel`
* Adds a PlantUML diagram for the new architecture (image above)

## Notes to reviewer
Call assignment functionality should be unchanged.

### Regarding in-store cache invalidation
In response to a comment I previously made on a PR by @ziggabyte (https://github.com/zetkin/app.zetkin.org/pull/845#discussion_r1031364707), where I wasn't feeling sure about where the logic to invalidate cache in response to "unrelated" actions should go, I have since reconsidered.

The models / repos / store separation of concerns in this PR is such:

* Models is the home of logic, even some logic whose main purpose is to render the correct UI. It does not contain cache logic (as previously imagined).
* Repositories and Store are both part of the "cache" layer
  * The repository contains the cache logic, to either retrieve from API or store. The repo is the only part of the system to communicate directly with the store.
  * The store contains the actual data and the reducer logic to assemble that data from incoming actions. It's part of the cache layer, so it makes sense for it to have a basic understanding of the caching logic, e.g. to invalidate things.

## Related issues
Resolves #847

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -43,7 +43,6 @@ module.exports = {
           },
         ],
         curly: 'error',
-        'brace-style': 'error',
```
Comment (ziggabyte) [inline .eslintrc.js:46]: CHAOS WILL ENSUE! 😂

File: src/core/api/client/BrowserApiClient.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    const res = await fetch(path, {
+      body: JSON.stringify(data),
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      method: 'PATCH',
+    });
+    const body = await res.json();
+    return body.data;
+  }
```
Comment (ziggabyte) [inline src/core/api/client/BrowserApiClient.ts:23]: so short and cute and clean!

File: src/core/caching/futures.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+export class RemoteItemFuture<DataType>
+  extends FutureBase<DataType>
+  implements IFuture<DataType>
+{
+  constructor(item: RemoteItem<DataType>) {
+    super(item.data, item.error, item.isLoading);
+  }
+}
```
Comment (ziggabyte) [inline src/core/caching/futures.ts:80]: i like the different futures

File: src/core/useModel.ts [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
+    throw new Error('Environment must be supplied. Add EnvProvider to tree.');
+  }
+
+  const [model, setModel] = useState(() => factory(env));
   const setRandom = useState(0)[1];
 
   useEffect(() => {
-    const unsubscribe = store.subscribe(() => {
+    const unsubscribe = env.store.subscribe(() => {
       setModel(() => model);
```
Comment (ziggabyte) [inline src/core/useModel.ts:23]: I understand this line is not part of this PR, but I have read it 10 times and i do not understand it and this is my note to ask you to help me understand this logic better.

Comment (ziggabyte) [reply src/core/useModel.ts:23]: (line 23, that is)

File: src/features/callAssignments/apiTypes.ts [hunk 5]
```diff
@@ -8,16 +8,15 @@ export interface CallAssignmentData {
...
   blocked: number;
   callBackLater: number;
   calledTooRecently: number;
   done: number;
-  isStale: boolean;
   missingPhoneNumber: number;
   organizerActionNeeded: number;
   queue: number;
   ready: number;
-  isLoading: boolean;
```
Comment (ziggabyte) [inline src/features/callAssignments/apiTypes.ts:22]: Yay for no more mixing!

File: src/features/callAssignments/models/CallAssignmentModel.spec.ts [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        missingPhoneNumber: 0,
+        organizerActionNeeded: 0,
+        queue: 0,
+        ready: 50,
+      });
+      expect(future.isLoading).toBeFalsy();
+      expect(future.error).toBeNull();
+    });
+  });
+});
```
Comment (ziggabyte) [inline src/features/callAssignments/models/CallAssignmentModel.spec.ts:244]: The structure in this test does go against the structure used in the Great Test Refactor, where we wanted to build test files that have only one "describe" clause, followed by "it" clauses. Does the testing of this new logic require more complex test structure? Or is it a habit, and can be re-structured?

File: src/features/callAssignments/store.ts [hunk 7]
```diff
@@ -1,98 +1,111 @@
...
+
       if (
-        //TODO: clean up this hot mess w better caching logic
-        callAssignment?.cooldown != action.payload.cooldown ||
-        JSON.stringify(action.payload.target.filter_spec) !=
-          JSON.stringify(callAssignment?.target.filter_spec)
+        statsItem &&
+        (callAssignment?.cooldown != action.payload.cooldown ||
+          JSON.stringify(action.payload.target.filter_spec) !=
+            JSON.stringify(callAssignment?.target.filter_spec))
```
Comment (ziggabyte) [inline src/features/callAssignments/store.ts:64]: rustique

**Review by ziggabyte — APPROVED** [review]
I approve, but look at my comment about the test and give your thoughts!

File: .eslintrc.js (hunk 1 above)
Comment (richardolsson) [reply .eslintrc.js:46]: I had to do this, because it turned out that prettier and eslint had conflicting ideas of how braces should be handled in some cases where classes are defined. By removing this, I let prettier decide, so no need to fear chaos. 😊

File: src/core/useModel.ts (hunk 4 above)
Comment (richardolsson) [reply src/core/useModel.ts:23]: It uses [functional state update](https://reactjs.org/docs/hooks-reference.html#functional-updates) but to be completely honest I'm not sure why I originally did it that way.

File: src/features/callAssignments/models/CallAssignmentModel.spec.ts (hunk 6 above)
Comment (richardolsson) [reply src/features/callAssignments/models/CallAssignmentModel.spec.ts:244]: This is a very valid point! I'm not sure I mind having multiple `describe()` when they very clearly relate to separate parts of the program code, like in this case where the other `describe()` is for the class, and then we have one for each tested piece of interface (property/method).

Let's chat about how it could be refactored after merging, if we want to. 💯
